### DISCUSSION TITLE
Running Code in an Isolated Process

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/runtime#100028] [Q&A]
Does .NET have any built-in (or easily accessible) way of running code in an isolated process?

Whether it is in .NET or a nuget package, I'm looking to be able to do something like this:
```

using var Host = await IsolatedHost.StartAsync(Bitness.x32);
var Instance = await Host.HostAsync<IMyClass, MyClass>();

await Instance.DoSomethingAsync();

```

I imagine the library using something like gRPC or JsonStreamRpc.

**Comment by tacosontitan** [comment]
```cs
var hostOne = Host.CreateDefaultBuilder(args).Build();
var hostTwo = Host.CreateDefaultBuilder(args).Build();
var hostingTasks = new List<Task>(capacity: 2)
{
    hostOne.RunAsync(...),
    hostTwo.RunAsync(...)
};

await Task.WhenAll(hostingTasks);
```

Would something like this not work?

<sub>Disclaimer... I'm not sure *why* you'd want to create separate hosts.</sub>

**Reply by TonyValenti** [reply]
No. That doesn't isolate the processes.

**Reply by tacosontitan** [reply]
Can you clarify "isolate the processes" then?

As my answer is written, you'd have two hosts in a single application capable of running their services independently of each other.

**Reply by TonyValenti** [reply]
An isolated process is a standalone process that does not share memory space with the main application. An isolated process can be used to overcome bitness constraints (a 64 bit process calls a 32 bit process which loads a 32 bit only DLL), elevation boundaries (think UAC prompts), and stability containment (imagine needing to call a DLL that is sometimes known to crash the process.  Crashing an isolated host won't crash your main exe).

**Comment by davidfowl** [comment]
https://github.com/tmds/Tmds.ExecFunction

**Comment by huoyaoyuan** [comment]
There are also remote executors used in runtime tests.
The complexity may come with operating system. It requires executable file and appropriate command to launch a process. In test suites, we often need to compile the new executable on the fly.

**Reply by davidfowl** [reply]
This is what ExecFunction is based on

**Reply by TonyValenti** [reply]
Thank!  This is really helpful.

For this app I'm working on, I will have a few different isolation hosts:
x32
x64
x64-elevated

This will require different EXEs to launch depending on what context I need the code to run in.

I have the following projects:
```
App.IsolatedHost (main code)
App.IsolatedHost._x32 (stub to App.IsolatedHost that will always build x86)
App.IsolatedHost._x64 (stub to App.IsolatedHost that will always build x64)
App.IsolatedHost._x64-Elevated (stub to App.IsolatedHost that will always build x64 with and elevated manifest).
```

I also have another project, ```App.IsolatedHosts`` and I want the following files in it
```
\App.IsolatedHosts\Resources\App.IsolatedHost._x32.exe
\App.IsolatedHosts\Resources\App.IsolatedHost._x64.exe
\App,IsolatedHosts\Resources\App.IsolatedHost._x64-Elevated.exe
```

Is there a way to configure my projects such that:
1.  ```App.IsolatedHost.*``` always build with SingleFile = true
2. The outputs of those builds are collected and saved into \App.IsolatedHosts\Resources\

My goal is that I can just build ```App.IsolatedHosts``` and it will build and package in all three of the process-specific hosts.
